### PR TITLE
Awd lm

### PR DESCRIPTION
**Pull request by cgraywang**
## Description ##

Add AWD-LSTM model for language model, add multi-gpu support for the model training, refine the data loader as the input for the model, and update related README for easily trying out.

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/gluon/model_zoo/text/base.py [hunk 1]
```diff
@@ -49,11 +49,18 @@ def forward(self, inputs, begin_state=None):
         encoded, state = self.encoder(embedded_inputs, begin_state)
         out = self.decoder(encoded)
         return out, state
+    
+    
+def _apply_weight_drop_to_rnn_cell(block, rate, weight_drop_mode = 'training'):
+    params = block.collect_params('h2h_weight')
+    weight_dropped_params = WeightDropParameter(params['h2h_weight'], rate, mode)
+    block.collect_params('h2h_weight').params['h2h_weight'] = weight_dropped_params
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/text/base.py]: This looks incorrect.

File: python/mxnet/gluon/model_zoo/text/base.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def forward(self, inputs, begin_state=None):
 def _apply_weight_drop_to_rnn_cell(block, rate, weight_drop_mode = 'training'):
     params = block.collect_params('h2h_weight')
     weight_dropped_params = WeightDropParameter(params['h2h_weight'], rate, mode)
-    block.collect_params('h2h_weight').params['h2h_weight'] = weight_dropped_params
+    block.collect_params('h2h_weight')._params['h2h_weight'] = weight_dropped_params
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/text/base.py]: what if the cell is created with a parameter dictionary as input to the constructor (i.e. `params=external_params_dict`? In that case the member that actual has the parameter would be in `_shared`.

**Comment by cgraywang** [comment]
@szha The rnn_cell has been tested and updated. Pls refer to the updated version.

File: python/mxnet/gluon/model_zoo/text/base.py [hunk 3]
```diff
@@ -49,22 +49,30 @@ def forward(self, inputs, begin_state=None):
...
-def get_rnn_cell(mode, num_layers, num_hidden,
+    
+    
+def _apply_weight_drop_to_rnn_cell(block, rate, weight_dropout_mode = 'training'):
+    params = block.collect_params('.*_h2h_weight')
+    for key, value in params.items():
+        weight_dropped_params = WeightDropParameter(value, rate, weight_dropout_mode)
+        block.collect_params('.*_h2h_weight')._params[key] = weight_dropped_params
+        for child_block in block._children:
+            child_block.collect_params('.*_h2h_weight')._params[key] = weight_dropped_params
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/text/base.py:60]: Call `_apply` recursively on children. It shouldn't be in the inner loop.
